ldclient_scoped.go

```diff
@@ -0,0 +1,621 @@
+package ldclient
+
+import (
+	gocontext "context"
+	"sync"
+
+	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
+	"github.com/launchdarkly/go-sdk-common/v3/ldmigration"
+	"github.com/launchdarkly/go-sdk-common/v3/ldreason"
+	"github.com/launchdarkly/go-sdk-common/v3/ldvalue"
+	ldevents "github.com/launchdarkly/go-sdk-events/v3"
+	"github.com/launchdarkly/go-server-sdk/v7/interfaces"
+)
+
+// LDScopedClient is a wrapper around LDClient that lets you specify the
+// evaluation context to be used for all operations, rather than taking the
+// evaluation context as a parameter every time you call a method.
+//
+// A LDScopedClient is created by calling [LDClient.ForContext]. This sets the
+// initial context to be used for all operations:
+//
+//	userContext := ldcontext.New("user-key")
+//	scopedClient := client.ForContext(userContext)
+//	scopedClient.BoolVariation("flag-key", false)
+//
+// Scoped contexts are mutable, to facilitate incrementally building up a
+// multi-context representing the current scope. You should create a new scoped
+// client for each logical scope where the contexts are different. For instance,
+// if you are using a scoped client in an HTTP request handler, you should create
+// a new scoped client for each request. You should only use the mutable features
+// of LDScopedClient to add new contexts, not to change or remove existing data.
+// A scoped client is thread-safe, so you can safely use it from multiple
+// goroutines.
+//
+// You may add new contexts with [LDScopedClient.AddContext]. The scoped client's
+// contexts so far are combined into a multi-context whenever the scoped client
+// is used. This is useful when more contexts become available later in the
+// lifecycle of a request. For instance, you might have a `user` context that is
```

Comment:
Optional suggestion to attempt to clarify that the scoped client only has one multi-context

```suggestion
// You may add new contexts to the scoped client's multi-context with [LDScopedClient.AddContext]. 
// This is useful when more contexts become available later in the
// lifecycle of a request. For instance, you might have a `user` context that is
```

Reply:
@mmrj Hmm, I think it isn't quite right to say that the scoped client contains a multi-context. It actually builds the multi-context on the fly when you need it. Also, multi-contexts are immutable, so talking about "adding contexts to the multi-context" may raise questions about whether you're actually changing the multi-context or just the client's reference to it (which it doesn't even have). However, I think you're right that we need some consistent way to refer to the "multi-context so far" that's implicitly available in a scoped client.

I've updated the comments in this file to use the term "current context" (which matches the method `CurrentContext()`) to refer the implicit inner multi-context, then use that consistently throughout when talking about adding stuff to the current context.

```
The "current context" is the
combination of all contexts added so far, as a multi-context.
```